docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,317 @@
+    namespaces of individual containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
+    initialization
+*   respect admission restrictions
+*   allow introspection of pod state (no hidden containers)
+*   support arbitrary runtimes via the CRI (possibly with reduced feature set)
+
+A good solution SHOULD:
+
+*   have an excellent user experience (i.e. should be a feature of the platform
+    rather than a config-time solution)
+*   require no direct access to the node
+*   have no inherent side effects to the running container image
+*   re-use existing container image distribution channels
+*   enable detection of pods that have been modified by an exec ("tainted" pods)
+
+## Implementation
+
+To troubleshoot a single container, one executes a process inside the container
+namespaces using `kubectl exec`. Since a pod is a collection of containers, it
+follows that to debug a pod we could execute a container inside the pod
+namespaces. This could be a logical extension of `kubectl exec` by adding a new
+flag to specify the name of the container image to run inside the pod namespace.
+
+The user would specify a debug image of her choice (e.g. `kubectl exec -it -m
+gcr.io/myproj/debugimage podname`) and be connected to a process running inside
+the pod. The processes of all other containers would be visible through the
+shared PID namespace. Network connections are visible through the shared net
+namespace. File systems are available via volume bind mounts, potentially
+read-only.
+
+### Prerequisites
+
+We'll have to complete several independent prerequisites to realize this
+functionality, including a few long standing feature requests. These would be
+fully described in separate proposals.
+
+#### Shared PID Namespace
+
+Tracked in [#1615](https://issues.k8s.io/1615), a shared PID namespace became
+available in Docker 1.12 and requires plumbing through the docker drivers,
+extending the infra container to reap orphaned zombies, and [rolling
+out](https://pr.k8s.io/37404).
+
+#### Adding a Container to a Pod Spec
+
+Tracked in [#37838](https://issues.k8s.io/37838), Kubernetes disallows almost
+all modification to a pod spec as a conservative default and because arbitrary
+changes could have complex repercussions on resource accounting and quality of
+service. The most straightforward way to add a container to a running pod is to
+modify the pod spec.
+
+We will enable a container to be added to the list of containers in a pod spec
+so long as the new container does not specify resource requirements. Removing a
+container is not allowed.
+
+A prohibition on changing resources simplifies scheduling, but it means a pod's
+QoS class could drop from *Guaranteed* to *Burstable*. This is an acceptable
+consequence of modifying a running pod. It means a pod could potentially be
+evicted during debugging, particularly if there are insufficient resources to
+allow for the troubleshooting process.
```

Comment:
I called out vertical resizing, but I couldn't find a proposal (and I didn't know about it prior to your comment).